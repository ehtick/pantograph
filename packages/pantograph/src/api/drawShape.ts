export { drawCircle } from "../drawShape/drawCircle.js";
export { drawRect } from "../drawShape/drawRect.js";
