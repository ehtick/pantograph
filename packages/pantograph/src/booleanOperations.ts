import {
  cutFiguresLists,
  fuseFiguresLists,
  intersectFiguresLists,
} from "./algorithms/boolean/figureBooleans";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { listOfFigures } from "./utils/listOfFigures";

export function fuse(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    fuseFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function fuseAll(shapes: (Diagram | Figure | Loop)[]): Diagram {
  return shapes.reduce(
    (acc: Diagram, shape: Diagram | Figure | Loop) => fuse(acc, shape),
    new Diagram()
  );
}

export function cut(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    cutFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}

export function intersect(
  first: Diagram | Figure | Loop,
  second: Diagram | Figure | Loop
): Diagram {
  return new Diagram(
    intersectFiguresLists(listOfFigures(first), listOfFigures(second))
  );
}
