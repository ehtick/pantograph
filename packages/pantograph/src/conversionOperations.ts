import { Segment } from "./models/segments/Segment";
import { Arc } from "./models/segments/Arc";
import { Line } from "./models/segments/Line";
import { QuadraticBezier } from "./models/segments/QuadraticBezier";
import { CubicBezier } from "./models/segments/CubicBezier";
import { EllipseArc } from "./models/segments/EllipseArc";
import { approximateBezierCurveAsArcs } from "./algorithms/conversions/bezierToArcs";
import { approximateEllipticalArcAsCubicBeziers } from "./algorithms/conversions/ellipseToBezier";
import { genericConversion } from "./algorithms/conversions/helpers";
import { Diagram, Figure, Loop, Strand } from "./main";

function mapAsArcsAndLines(s: Segment, tolerance = 1e-3): Segment[] {
  if (s instanceof Arc || s instanceof Line) {
    return [s];
  }
  if (s instanceof QuadraticBezier || s instanceof CubicBezier) {
    return approximateBezierCurveAsArcs(s, tolerance);
  }
  if (s instanceof EllipseArc) {
    return approximateEllipticalArcAsCubicBeziers(s).flatMap((s) =>
      approximateBezierCurveAsArcs(s, tolerance),
    );
  }

  throw new Error(`Unknown segment type: ${s}`);
}

function convertToArcsAndLines(shape: Diagram): Diagram;
function convertToArcsAndLines(shape: Figure): Figure;
function convertToArcsAndLines(shape: Loop): Loop;
function convertToArcsAndLines(shape: Strand): Strand;
function convertToArcsAndLines(shape: Segment): Segment[];
function convertToArcsAndLines(
  shape: Diagram | Figure | Loop | Strand | Segment,
  tolerance = 1e-3,
) {
  const mapper = (segment: Segment) => mapAsArcsAndLines(segment, tolerance);
  return genericConversion(shape, mapper);
}

export { convertToArcsAndLines };
