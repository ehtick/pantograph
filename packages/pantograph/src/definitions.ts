export type Vector = [number, number];

export function isVector(vector: unknown): vector is Vector {
  return Array.isArray(vector) && vector.length === 2;
}
