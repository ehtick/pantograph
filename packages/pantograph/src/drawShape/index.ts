export { drawCircle } from "./drawCircle.js";
export { drawRect } from "./drawRect.js";
export { drawEllipse } from "./drawEllipse.js";
export { drawSVGPath } from "./drawSVGPath.js";
