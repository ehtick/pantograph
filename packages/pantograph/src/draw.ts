import { Vector } from "./definitions";
import { Diagram } from "./models/Diagram";
import { Figure } from "./models/Figure";
import { Loop } from "./models/Loop";
import { Line } from "./models/segments/Line";
import { Segment } from "./models/segments/Segment";
import { TransformationMatrix } from "./models/TransformationMatrix";
import {
  polarToCartesian,
  DEG2RAD,
  subtract,
  sameVector,
} from "./vectorOperations";

function closeSegments(segments: Segment[]) {
  if (!segments.length) throw new Error("No segments to close");
  const firstSegment = segments[0];
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const lastSegment = segments.at(-1)!;

  if (sameVector(firstSegment.firstPoint, lastSegment.lastPoint))
    return segments;
  return [
    ...segments,
    new Line(lastSegment.lastPoint, firstSegment.firstPoint),
  ];
}

function loopySegmentsToDiagram(segments: Segment[]) {
  // Here we will need to do our best to fix cases where the drawing is
  // broken in some way (i.e. self-intersecting loops)

  return new Diagram([new Figure(new Loop(segments))]);
}

export class DrawingPen {
  protected pointer: Vector;
  protected firstPoint: Vector;
  protected pendingSegments: Segment[];

  constructor(origin: Vector = [0, 0]) {
    this.pointer = origin;
    this.firstPoint = origin;

    this.pendingSegments = [];
  }

  movePointerTo(point: Vector): this {
    if (this.pendingSegments.length)
      throw new Error(
        "You can only move the pointer if there is no segment defined"
      );

    this.pointer = point;
    this.firstPoint = point;
    return this;
  }

  protected saveSegment(segment: Segment) {
    this.pendingSegments.push(segment);
    return this;
  }

  lineTo(point: Vector): this {
    const segment = new Line(this.pointer, point);
    this.pointer = point;
    return this.saveSegment(segment);
  }

  line(xDist: number, yDist: number): this {
    return this.lineTo([this.pointer[0] + xDist, this.pointer[1] + yDist]);
  }

  vLine(distance: number): this {
    return this.line(0, distance);
  }

  hLine(distance: number): this {
    return this.line(distance, 0);
  }

  vLineTo(yPos: number): this {
    return this.lineTo([this.pointer[0], yPos]);
  }

  hLineTo(xPos: number): this {
    return this.lineTo([xPos, this.pointer[1]]);
  }

  polarLineTo([r, theta]: Vector): this {
    const angleInRads = theta * DEG2RAD;
    const point = polarToCartesian(r, angleInRads);
    return this.lineTo(point);
  }

  polarLine(distance: number, angle: number): this {
    const angleInRads = angle * DEG2RAD;
    const [x, y] = polarToCartesian(distance, angleInRads);
    return this.line(x, y);
  }

  tangentLine(distance: number): this {
    const previousCurve = this.pendingSegments.at(-1);

    if (!previousCurve)
      throw new Error("You need a previous segment to sketch a tangent line");

    const [xDir, yDir] = previousCurve.tangentAtLastPoint;
    return this.line(xDir * distance, yDir * distance);
  }

  close(): Diagram {
    const segments = closeSegments(this.pendingSegments);
    return loopySegmentsToDiagram(segments);
  }

  closeWithMirror(): Diagram {
    if (!this.pendingSegments.length) throw new Error("No segments to close");

    const firstSegment = this.pendingSegments[0];
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const lastSegment = this.pendingSegments.at(-1)!;

    const mirrorVector = subtract(
      lastSegment.lastPoint,
      firstSegment.firstPoint
    );
    const mirrorTranform = new TransformationMatrix().mirrorLine(
      mirrorVector,
      firstSegment.firstPoint
    );

    const mirroredSegments = this.pendingSegments.map((segment) =>
      segment.transform(mirrorTranform).reverse()
    );
    mirroredSegments.reverse();

    return loopySegmentsToDiagram([
      ...this.pendingSegments,
      ...mirroredSegments,
    ]);
  }
}

export function draw(origin: Vector = [0, 0]): DrawingPen {
  return new DrawingPen(origin);
}
