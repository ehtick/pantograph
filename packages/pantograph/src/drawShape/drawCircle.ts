import { draw } from "../draw";
import type { Diagram } from "../models/Diagram";

export function drawCircle(radius: number): Diagram {
  return draw([-radius, 0])
    .sagittaArc(2 * radius, 0, radius)
    .sagittaArc(-2 * radius, 0, radius)
    .close();
}
