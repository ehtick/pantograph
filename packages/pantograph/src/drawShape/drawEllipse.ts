import { draw } from "../draw.js";
import type { Diagram } from "../main.js";

export function drawEllipse(rx: number, ry: number): Diagram {
  const pen = draw([-rx, 0]);
  pen.halfEllipse(2 * rx, 0, ry);
  pen.halfEllipse(-2 * rx, 0, ry);
  return pen.close();
}
