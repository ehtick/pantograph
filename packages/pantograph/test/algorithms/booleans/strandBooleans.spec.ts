import { describe, it, expect } from "vitest";

import { draw } from "../../../src/draw";
import { drawRect } from "../../../src/drawShape";

import { debugImg } from "../../debug";

import {
  eraseStrandWithinLoop,
  eraseStrandOutsideLoop,
} from "../../../src/algorithms/boolean/strandBoolean";

// Note: all the tests are checked visually by using `debugImg`. Please do so
// as well when you add new tests or have some failures.

describe("eraseStrandWithinLoop", () => {
  it("should erase the parts of a strand within a loop", () => {
    const r = drawRect(5, 3).figures[0].contour;
    const s = draw([-3, -3])
      .lineTo([1, -1])
      .lineTo([0.5, 1])
      .lineTo([3, 3])
      .asStrand();

    expect(eraseStrandWithinLoop(s, r)).toMatchSnapshot();
  });

  it("should not erase the parts of on the border when configured that way", () => {
    const r = drawRect(5, 3).figures[0].contour;
    const s = draw([-3, -3])
      .lineTo([1, -1.5])
      .lineTo([-1, -1.5])
      .lineTo([0.5, 1])
      .lineTo([3, 3])
      .asStrand();

    expect(eraseStrandWithinLoop(s, r)).toMatchSnapshot();
  });

  it("should erase the parts of on the border when configured that way", () => {
    const r = drawRect(5, 3).figures[0].contour;
    const s = draw([-3, -3])
      .lineTo([1, -1.5])
      .lineTo([-1, -1.5])
      .lineTo([0.5, 1])
      .lineTo([3, 3])
      .asStrand();

    expect(eraseStrandWithinLoop(s, r, true)).toMatchSnapshot();
  });
});

describe("eraseStrandOutsideLoop", () => {
  it("should erase the parts of a strand within a loop", () => {
    const r = drawRect(5, 3).figures[0].contour;
    const s = draw([-3, -3])
      .lineTo([1, -1])
      .lineTo([0.5, 1])
      .lineTo([3, 3])
      .asStrand();

    expect(eraseStrandOutsideLoop(s, r)).toMatchSnapshot();
  });

  it("should not erase the parts of on the border when configured that way", () => {
    const r = drawRect(5, 3).figures[0].contour;
    const s = draw([-3, -3])
      .lineTo([1, -1.5])
      .lineTo([-1, -1.5])
      .lineTo([0.5, 1])
      .lineTo([3, 3])
      .asStrand();

    expect(eraseStrandOutsideLoop(s, r)).toMatchSnapshot();
  });

  it("should erase the parts of on the border when configured that way", () => {
    const r = drawRect(5, 3).figures[0].contour;
    const s = draw([-3, -3])
      .lineTo([1, -1.5])
      .lineTo([-1, -1.5])
      .lineTo([0.5, 1])
      .lineTo([3, 3])
      .asStrand();

    debugImg([r, s]);
    debugImg(eraseStrandOutsideLoop(s, r, true), "erased");

    expect(eraseStrandOutsideLoop(s, r, true)).toMatchSnapshot();
  });
});
