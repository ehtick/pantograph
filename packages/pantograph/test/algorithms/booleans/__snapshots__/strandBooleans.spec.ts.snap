// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`eraseStrandOutsideLoop > should erase the parts of a strand within a loop 1`] = `
[
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          1,
          0.5,
        ],
        "_boundingBox": null,
        "firstPoint": [
          0,
          -1.5,
        ],
        "lastPoint": [
          1,
          -1,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
      Line {
        "_V": [
          -0.5,
          2,
        ],
        "_boundingBox": null,
        "firstPoint": [
          1,
          -1,
        ],
        "lastPoint": [
          0.5,
          1,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
      Line {
        "_V": null,
        "_boundingBox": null,
        "firstPoint": [
          0.5,
          1,
        ],
        "lastPoint": [
          1.125,
          1.5,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
]
`;

exports[`eraseStrandOutsideLoop > should erase the parts of on the border when configured that way 1`] = `
[
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          1.5,
          2.5,
        ],
        "_boundingBox": null,
        "firstPoint": [
          -1,
          -1.5,
        ],
        "lastPoint": [
          0.5,
          1,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
      Line {
        "_V": null,
        "_boundingBox": null,
        "firstPoint": [
          0.5,
          1,
        ],
        "lastPoint": [
          1.125,
          1.5,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
]
`;

exports[`eraseStrandOutsideLoop > should not erase the parts of on the border when configured that way 1`] = `
[
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          -2,
          0,
        ],
        "_boundingBox": null,
        "firstPoint": [
          1,
          -1.5,
        ],
        "lastPoint": [
          -1,
          -1.5,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          1.5,
          2.5,
        ],
        "_boundingBox": null,
        "firstPoint": [
          -1,
          -1.5,
        ],
        "lastPoint": [
          0.5,
          1,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
      Line {
        "_V": null,
        "_boundingBox": null,
        "firstPoint": [
          0.5,
          1,
        ],
        "lastPoint": [
          1.125,
          1.5,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
]
`;

exports[`eraseStrandWithinLoop > should erase the parts of a strand within a loop 1`] = `
[
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          3,
          1.5,
        ],
        "_boundingBox": null,
        "firstPoint": [
          -3,
          -3,
        ],
        "lastPoint": [
          0,
          -1.5,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          1.875,
          1.5,
        ],
        "_boundingBox": null,
        "firstPoint": [
          1.125,
          1.5,
        ],
        "lastPoint": [
          3,
          3,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
]
`;

exports[`eraseStrandWithinLoop > should erase the parts of on the border when configured that way 1`] = `
[
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          4,
          1.5,
        ],
        "_boundingBox": null,
        "firstPoint": [
          -3,
          -3,
        ],
        "lastPoint": [
          1,
          -1.5,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          1.875,
          1.5,
        ],
        "_boundingBox": null,
        "firstPoint": [
          1.125,
          1.5,
        ],
        "lastPoint": [
          3,
          3,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
]
`;

exports[`eraseStrandWithinLoop > should not erase the parts of on the border when configured that way 1`] = `
[
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          4,
          1.5,
        ],
        "_boundingBox": null,
        "firstPoint": [
          -3,
          -3,
        ],
        "lastPoint": [
          1,
          -1.5,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          -2,
          0,
        ],
        "_boundingBox": null,
        "firstPoint": [
          1,
          -1.5,
        ],
        "lastPoint": [
          -1,
          -1.5,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
  Strand {
    "_boundingBox": null,
    "segments": [
      Line {
        "_V": [
          1.875,
          1.5,
        ],
        "_boundingBox": null,
        "firstPoint": [
          1.125,
          1.5,
        ],
        "lastPoint": [
          3,
          3,
        ],
        "precision": 1e-9,
        "segmentType": "LINE",
      },
    ],
    "strokeType": "STRAND",
  },
]
`;
